src/Microsoft.Extensions.Configuration.Abstractions/IConfigurationBuilder.cs

```diff
@@ -26,11 +26,11 @@ public interface IConfigurationBuilder
         /// </summary>
         /// <param name="source">The configuration source to add.</param>
         /// <returns>The same <see cref="IConfigurationBuilder"/>.</returns>
-        IConfigurationBuilder Add(IConfigurationSource source);
+        new IConfigurationBuilder Add(IConfigurationSource source);
```

Comment:
@davidfowl 

The change as proposed in this PR only made sense if we get rid of `Sources` completely. If you want encapsulation we should consider just adding the `Clear` method as you proposed, but:

1. We are already exposing properties as a dictionary. 
2. One can simply cast the `IEnumerable` returned by `Sources` to a list and play with internal data.

Not saying it is a good thing, but there isn't much encapsulation left to break.